### ISSUE TITLE
`ddev import-db` fails with "Error: Unknown command '\-'" because of new directive in mariadb-dump output

### ISSUE THREAD

**Issue by ioanma** [ddev/ddev#6249]
## TL;DR:

If  `ddev import-db` or `ddev pull` is failing with `Error: Unknown command '\-'"`

* Make sure you're using DDEV v1.23.1+
* Update the appropriate ddev-dbserver image based on your version of Mariadb in the .ddev/config.yaml
```
docker pull ddev/ddev-dbserver-mariadb-10.11:v1.23.1
docker pull ddev/ddev-dbserver-mariadb-10.6:v1.23.1
docker pull ddev/ddev-dbserver-mariadb-10.5:v1.23.1
```

If you are not using one of those mariadb versions (10.11, 10.6, 10.5) you'll need to migrate to one as one to get the newer version of `mysql` in those versions. 

### Expected Behavior

Uninterrupted db file import

### Actual Behavior

```
Error:

ddev drush sql-cli < dump_file.sql
> ERROR at line 1: Unknown command '\-'.

In Process.php line 270:

  The command "mysql --defaults-file=/tmp/drush_Bh6glU --database=db --host=db --port=3306 -A" failed.

  Exit Code: 1(General error)

  Working directory: /var/www/html/web

  Output:
  ================


  Error Output:
  ================
  ERROR at line 1: Unknown command '\-'.
```

### Steps To Reproduce

Import db dump (from older mysql client) using:

`ddev import-db src={PATH_TO_DB_FILE}`

or

`ddev drush sql-cli {PATH_TO_DB_FILE}`

For older mysql clients, the following line is inserted in the db dump file, causing the above error to occur.

```
/*!999999\- enable the sandbox mode */ 
```

### Anything else?

The issue is described here: https://mariadb.org/mariadb-dump-file-compatibility-change/

**Comment by rfay** [comment]
Hi @ioanma - MariaDB last week changed their dump format in a patch version of the product. 

This is discussed a number of places, especially
* https://github.com/ddev/ddev/issues/6083#issuecomment-2116597297
* https://mariadb.org/mariadb-dump-file-compatibility-change/
* https://stackoverflow.com/questions/78530098/ddev-import-db-error-unknown-command

The bottom line is that the problem is in your sql dump, you'll have to use a workaround for now, given in one of those places. 

It's quite surprising and concerning that a vendor would make a massive and incompatible change like this in a patch version of the product. 

I'll leave this issue open as it will have a better title the the other discussion.

**Comment by ioanma** [comment]
That makes sense @rfay and thanks for organising this (apologies I should've searched more before creating a new issue).

**Comment by rfay** [comment]
Well, you searched enough to come up with the answer, and that's a lot! Since everybody will hit this if using newer MariaDB we need it to be obvious. It's possible that we'll be able to add a workaround in next DDEV version.

**Comment by ioanma** [comment]
To be honest that's what I was thinking when opening the issue: Check whether this line (or a pattern of it in case there are other variations possible) is present and remove it before importing the db file. I understand that it's not ddev's responsibility to clean MariaDB's mess but it'd still be helpful as a quick and silent fix.

**Comment by 24bitIO** [comment]
Hello.
If the 10.11 docker image were updated to 10.11.8 rather than 10.11.7, wouldn't this work?
Forgive me if this is a stupid question.

https://mariadb.org/mariadb-dump-file-compatibility-change/

> Affected Versions
> The change is happening / has happened in MariaDB 10.5.25, 10.6.18, 10.11.8, 11.0.6, 11.1.5, 11.2.4 and 11.4.2. So, if you take a dump from these versions and try and import it into an unsupported version, such as 10.4, or an older bugfix version such as 10.5.24, you would hit this problem.

**Comment by rfay** [comment]
> If the 10.11 docker image were updated to 10.11.8 rather than 10.11.7, wouldn't this work?

Yes, in v1.23.2, the mariadb 10.11 image will be updated to 10.11.x (whatever is current)

That will solve the problem for `ddev import-db` for mariadb:10.11, for new database dumps being imported into 10.11. 

It won't likely solve all problems with `drush sql*` actions, which depend on mysql/mysqldump in the web container, which usually has mysqldump package from a different source.  That needs 
* https://github.com/ddev/ddev/issues/6083

And it likely won't solve all the million other permutations of a dump coming from an upstream database server which is not on the exact same patch level, although I'm hoping that the new mariadb-dump will happily handle older-style dumps. 

For 20 years there was relative stability in this, and then they broke it in a patch release!

**Comment by hussainweb** [comment]
This is one of the reasons I started looking into MariaDB 11.4 compatibility in #6243. The problem is much weirder though.

I know that this problem will be fixed (at least for `db-import` with `v1.23.2`. But until then, would you consider updating the images at `v1.23.1` so that it uses the latest patch releases of the database images? I understand that it can't fix all the problems outlined in this thread but this will fix some of them. I understand if you have a policy of not updating tagged images once pushed but I thought I'd ask, just in case.

If this is not possible, can you share how far away is the `v1.23.2` release? I don't mean to push but our CI builds are failing and based on this, we can plan alternatives.

**Comment by hussainweb** [comment]
Thanks! Just to clarify, if you can overwrite the database images, I am not concerned about 1.23.2 release then. :)

**Comment by stasadev** [comment]
I have pushed the affected v1.23.1 images:

Depending on the workflow, people need to pull the images again:

```
docker pull ddev/ddev-dbserver-mariadb-10.11:v1.23.1
```

```
docker pull ddev/ddev-dbserver-mariadb-10.6:v1.23.1
```

```
docker pull ddev/ddev-dbserver-mariadb-10.5:v1.23.1
```

```
$ docker run -it --rm --entrypoint=bash ddev/ddev-dbserver-mariadb-10.11:v1.23.1
$ mariadb -V
mariadb  Ver 15.1 Distrib 10.11.8-MariaDB, for debian-linux-gnu (x86_64) using  EditLine wrapper

$ docker run -it --rm --entrypoint=bash ddev/ddev-dbserver-mariadb-10.6:v1.23.1
$ mariadb -V
mariadb  Ver 15.1 Distrib 10.6.18-MariaDB, for debian-linux-gnu (x86_64) using readline 5.2

$ docker run -it --rm --entrypoint=bash ddev/ddev-dbserver-mariadb-10.5:v1.23.1
$ mariadb -V
mariadb  Ver 15.1 Distrib 10.5.25-MariaDB, for debian-linux-gnu (x86_64) using readline 5.2
```

**Comment by rfay** [comment]
Please let us know if this sorts out some of the problems for you.

**Comment by hussainweb** [comment]
@rfay, yes, my CI pipelines are working again (I made sure to clear the Docker images cache). Thanks for pushing this so quickly.

**Comment by rfay** [comment]
@hussainweb I can't remember, is it you who is using platform.sh and their upgrade failed it? I'm unable to get a bad dump out of Platform.sh at this point.

**Comment by hussainweb** [comment]
I don't think I mentioned platform.sh but yes, that was the reason for me. Initially, I thought it was because I was on MariaDB 11.0 on platform.sh and 10.11 on DDEV (hence the other PR). But I later found out that the problem was to do with dump file breaking change as described in this issue. If you're interested, [this is the action run](https://github.com/contrib-tracker/backend/actions/runs/9227613994/job/25390495576#step:10:64). I ran the same job again after clearing caches and [it worked](https://github.com/contrib-tracker/backend/actions/runs/9227613994/job/25473127094).

**Comment by rfay** [comment]
@hussainweb Could you please PM me your setup (.platform/services.yaml)? Maybe switch to PM in Discord? I changed our automated test environment to 10.11 and got 10.11.8 on server side there, but I don't get broken new-style db dumps for testing from it because it's still using the mariadb 10.3 version of mysqldump.

**Comment by hussainweb** [comment]
@rfay, I don't mind sharing in PM but the project in question is open-source. You can see the file here: https://github.com/contrib-tracker/backend/blob/main/.platform/services.yaml

**Comment by rfay** [comment]
I confirm that with current mariadb:10.11 images
* `ddev export-db` exports with the new-style header in the dump
* `ddev import-db` imports the new-style header just fine. 
* `ddev import-db` with the new image does fine with old-style mysql dumps.

I also confirm that, as expected, `drush sql-cli` can't work with the new header. So inside `ddev ssh`

```
$ gzip -dc .tarballs/db.sql.gz | drush sql-cli
> ERROR at line 1: Unknown command '\-'.
```

**Comment by MakerTim** [comment]
For the work arround

We currently add a command in the providers/ENV.yml that will remove the broken line.
MariaDB suggests to `tail +2` the file, but with `drush sql:dump` and other stacked dumps they will fail on the next occurrence.
So we do
```sed -i '/^\/\*!999999\\- enable the sandbox mode \*\//d' ${DDEV_APPROOT}/.ddev/.downloads/db.sql```
This will remove all occurrences in the db-file locally just before import

**Comment by rfay** [comment]
Now that #6253 has landed, DDEV HEAD `ddev import-db` and related things like `ddev pull` will automatically remove the ugly new line from the dumpfile when loading it.

**Comment by jiansong** [comment]
> I have pushed the affected v1.23.1 images:
>  
>  [...]

The `ddev/ddev-webserver` image also needs to be patched, because it still has the affected client version. 

For example, command like `ddev exec php <framework_cli> db:import` is broken right now, because it's run inside the webserver container.

**Comment by rfay** [comment]
@jiansong this class of problem is discussed in
* https://github.com/ddev/ddev/issues/6083#issuecomment-2116597297

If you use `ddev import-db` or related commands things should be working already. But AFAICT the packages are not yet ready for use on generic Debian, but we'll be trying to get a mysql client into the web container in v1.23.2 that can handle the new import format. Please follow #6083.

In the meantime you can preprocess the database dump as described several places.

**Comment by rfay** [comment]
I'd appreciate any reviews of this blog post: https://github.com/ddev/ddev.com/pull/210 - the rendered version is at https://20240613-mariadb.ddev-com-front-end.pages.dev/blog/mariadb-dump-breaking-change/

**Comment by LenLamberg** [comment]
I encountered this problem when I used ssh to export the database from the remote site and then imported it using DDEV on my local site.  I received the error:  ERROR at line 1: Unknown command '\-'   I was able to work around the error by opening the SQL dump file in a text editor and deleting the first line, which is "/*!999999\- enable the sandbox mode */ "

**Comment by rfay** [comment]
@LenLamberg were you using current DDEV v1.23.3 and doing `ddev import-db`?  You can use `ddev --version` to see what version you have.

**Comment by LenLamberg** [comment]
Mr. Fay, nice to hear from you!  We spoke once (virtually, I think) at a MidCamp a few years ago.  You are an evangelist for DDEV

My current version is v1.21.1

I regularly update ubuntu through sudo apt update and then sudo apt upgrade.  This does not seem to update my ddev.  I checked location by going which -a ddev, and I have a single installation at /usr/local/bin/ddev

As I explained previously, the import fails when I simply use ddev import-db (dump file name).sql  However, it succeeds when I remove the first line of the dump file.

Thanks again for reaching out to me

Len Lamberg, ***@***.***


------ Original Message ------
From "Randy Fay" ***@***.******@***.***>>
To "ddev/ddev" ***@***.******@***.***>>
Cc "leonard lenlamberg.com" ***@***.******@***.***>>; "Mention" ***@***.******@***.***>>
Date 7/5/2024 10:56:20 PM
Subject Re: [ddev/ddev] `ddev import-db` fails with "Error: Unknown command '\-'" because of new directive in mariadb-dump output (Issue #6249)


@LenLamberg<https://github.com/LenLamberg> were you using current DDEV v1.23.3 and doing ddev import-db? You can use ddev --version to see what version you have.

**Comment by rfay** [comment]
Hi @LenLamberg - Please upgrade to v1.23.3 to solve this problem. You can tell when you have upgrade when `ddev --version` says `v1.23.3`.  You can read about what happened at https://ddev.com/blog/mariadb-dump-breaking-change/ - which is all explained in the issue.

**Comment by rfay** [comment]
Here's the doc: https://ddev.readthedocs.io/en/stable/users/install/ddev-upgrade/

I imagine that you didn't install DDEV via apt. You can remove whatever you currently have as described there. Consider `brew uninstall ddev` and `rm $(which ddev)`.  Then install it with apt.

**Comment by LenLamberg** [comment]
Thank you again, the problem is solved: When DDEV is updated to v1.23.3, the import works without any problem.  I had some difficulty updating DDEV: it wasn't as simple as sudo apt upgrade, which did not update the install at /usr/local/bin/ddev.  I removed the ddev file at that location and then re-installed DDEV using the "WSL2 + Docker CE Inside Install Script" found at https://ddev.readthedocs.io/en/stable/users/install/ddev-installation/#windows  This solution is secure, and I no longer need to run Docker Desktop.
